python/lsst/ts/tcpip/message_reading_one_client_server.py

```diff
@@ -0,0 +1,151 @@
+# This file is part of ts_tcpip.
+#
+# Developed for the Rubin Observatory Telescope and Site System.
+# This product includes software developed by the LSST Project
+# (https://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+import abc
+import asyncio
+import logging
+import typing
+
+from .base_client_or_server import BaseClientOrServer, ConnectCallbackType
+from .constants import DEFAULT_MONITOR_CONNECTION_INTERVAL, TERMINATOR
+from .one_client_server import OneClientServer
+
+__all__ = ["MessageReadingOneClientServer"]
+
+
+class MessageReadingOneClientServer(OneClientServer):
```

Comment:
Speaking of all those extra write methods...it is a misfeature that this class offers public read and write methods that subclasses should not touch. This is especially an issue for the read methods, because calling one of those will confuse the read loop.

I'm not sure if it's worth doing anything about it, but I think it is important to think about that before we adopt this code, because it will be much harder to fix later.

My favorite solution I have thought of, so far, is:
* Add an abstract read_loop method to OneClientServer and change all existing users to use it/
* Make the read methods private in OneClientServer (or at least documented as "don't touch"), with the expectation that the concrete read_loop will pick the right one
Advantages include:
* It eliminates the need for a background connection monitoring task.
* All servers have read loops (unless they are write-only, in which case the read loop can be trivial) so it is a natural fit.

But it will need some backwards- incompatible changes, especially if we make the read methods private. I am perfectly willing to fix my code if you are interested. But I should do a few experimental conversions to be sure this makes sense.

Another option is to make this class contain a server instead of act as a server. The problem with this is that this class *does* act as a server and think it will be confusing that it contains a server instead of acting as one (certainly I'd like a name that does not have OneClientServer in it). Also, enough of the OneClientServer methods are applicable that it will need some trivial shadow methods (e.g. a version of close that calls self.server.close).

An option I am not keen on is to make a base class that has private read and write methods and make all of the public classes inherit from that. This hides public methods, but requires even more shadow methods than the previous suggestion, especially in OneClientServer, which would need to shadow all the read and write methods.

Another option is to ask for advice on the software-dev slack channel. This is reaching the limits of my software design skills and sometimes I have gotten very good advice there.